src/poetry/repositories/pypi_repository.py

```diff
@@ -453,3 +454,40 @@ def _download(self, url: str, dest: str) -> None:
 
     def _log(self, msg: str, level: str = "info") -> None:
         getattr(logger, level)(f"<debug>{self._name}:</debug> {msg}")
+
+    def get_sha_hash_from_link(self, link: Link) -> str | None:
+        """Get sha256|384|512 hash for a file from the provided link.
+
+        If the hash type included in the link is not sha256|384|512,
+        convert it to sha256.
+
+        """
+        file_hash = f"{link.hash_name}:{link.hash}" if link.hash else None
+
+        if not link.hash or (
+            link.hash_name not in ("sha256", "sha384", "sha512")
+            and hasattr(hashlib, link.hash_name)
+        ):
+            with temporary_directory() as temp_dir:
+                filepath = Path(temp_dir) / link.filename
+                self._download(link.url, str(filepath))
```

Comment:
This might get a bit problematic as it could end up folks downloading all available builds for all dependencies, well n+1. In the interim I reckon best to download to cache and use the downloaded file as is without moving it around. Atleast that will save redownload when relocking or re-solving.

Reply:
Reviewing the codebase, it looks like I'm going to have to pass `poetry.installation.chef.Chef` into the repository object or create it within the object to have access to the caching logic. To do so, I would need to know the `Env` when registering the repos (at a minimum when registering the legacy repos). I'm having trouble locating where I could pull `Env` from at that point. Any advice?

The other option I can think of would to be pass `Chef` as an argument to `Repository.get_release_info`/`Repository.package` & `Chooser.choose_for` since all of them would need to have the ability to download and calculate sha256 for a file. But, that seems excessive when it's only needed in cerin circumstances.